File: src/VBNetFx/FacGenericControls/ucAnalysisDateRangePicker.vb
Imports ActiveDev
Imports Facesso.Data
Imports System.ComponentModel

Public Class ucAnalysisDateRangePicker

    Private myDateRangeValue As DateRangeParameter
    Private myDoNothing As Boolean
    Private myLastWorkingday As LastWorkingdays

    Public Sub New()
        ' This call is required by the Windows Form Designer.
        myDoNothing = True
        InitializeComponent()
        myDoNothing = False
        myLastWorkingday = LastWorkingdays.Friday

        ' Add any initialization after the InitializeComponent() call.
        cmbMonthsHistory.Items.Add("Laufender Monat (" & MonthText(0) & ")")
        cmbMonthsHistory.Items.Add("Letzter Monat (" & MonthText(1) & ")")
        For c As Integer = 2 To 24
            cmbMonthsHistory.Items.Add("Vor " & c & " Monaten (" & MonthText(c) & ")")
        Next
        DateRangeValue = New DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, 1)
        myDateRangeValue.LastWorkingday = LastWorkingdays.Friday

    End Sub

    Private Function MonthText(ByVal MonthsIntoPast As Integer) As String
        Dim locStartDate As Date
        locStartDate = Dates.LastDayOfMonth(Date.Now).AddMonths(-MonthsIntoPast)
        Return locStartDate.ToString("MMM, yyyy")
    End Function

    Public Property DateRangeValue() As DateRangeParameter
        Get
            Return myDateRangeValue
        End Get
        Set(ByVal value As DateRangeParameter)
            myDateRangeValue = value
            UpdateUI()
        End Set
    End Property

    Public Property LastWorkingday() As LastWorkingdays
        Get
            Return myLastWorkingday
        End Get
        Set(ByVal value As LastWorkingdays)
            myLastWorkingday = value
            myDateRangeValue.LastWorkingday = value
        End Set
    End Property

    Private Sub UpdateUI()
        myDoNothing = True
        Select Case myDateRangeValue.DateRangePreset
            Case DateRangePresets.CustomPeriod
                optCustomPeriod.Checked = True
                myDoNothing = False
                Return
            Case DateRangePresets.YesterdayOrLastWorkingDay
                optYesterday.Checked = True
            Case DateRangePresets.FromStartOfCurrentMonthToNow
                optFromStartOfCurrentMonthToNow.Checked = True
            Case DateRangePresets.FromStartOfCurrentWeekToNow
                optFromStartOfCurrentWeekToNow.Checked = True
            Case DateRangePresets.FromStartToEndOfSpecifiedMonth
                cmbMonthsHistory.SelectedIndex = DateRangeValue.MonthIntoPast
            Case DateRangePresets.LastWeek
                optLastWeek.Checked = True
            Case DateRangePresets.Today
                optToday.Checked = True
            Case DateRangePresets.SinceYearBeganToNow
                optSinceYearBegan.Checked = True
            Case Else
                optWeekBeforeLastWeek.Checked = True
        End Select
        dtpStart.Value = myDateRangeValue.StartDate
        dtpEnd.Value = myDateRangeValue.EndDate
        myDoNothing = False
    End Sub

    Private Sub dtpStart_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles dtpStart.ValueChanged
        If myDoNothing Then
            Return
        End If
        If dtpStart.Value > dtpEnd.Value Then
            dtpEnd.Value = dtpStart.Value
        End If
        DateRangeValue = New DateRangeParameter(dtpStart.Value, dtpEnd.Value)
    End Sub

    Private Sub dtpEnd_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles dtpEnd.ValueChanged
        If myDoNothing Then
            Return
        End If
        If dtpStart.Value > dtpEnd.Value Then
            dtpStart.Value = dtpEnd.Value
        End If
        DateRangeValue = New DateRangeParameter(dtpStart.Value, dtpEnd.Value)
    End Sub

    <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
     EditorBrowsable(EditorBrowsableState.Always)> _
    Public Overrides Property Text() As String
        Get
            Return MyBase.Text
        End Get
        Set(ByVal value As String)
            MyBase.Text = value
        End Set
    End Property

    Protected Overrides Sub OnTextChanged(ByVal e As System.EventArgs)
        MyBase.OnTextChanged(e)
        gbTitle.Text = Me.Text
    End Sub

    Private Sub optDateRanges_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) _
        Handles optYesterday.CheckedChanged, optFromStartOfCurrentMonthToNow.CheckedChanged, _
        optFromStartOfCurrentWeekToNow.CheckedChanged, optStartToEndOfSpecifiedMonth.CheckedChanged, _
         optLastWeek.CheckedChanged, _
        optToday.CheckedChanged, optSinceYearBegan.CheckedChanged, _
        optWeekBeforeLastWeek.CheckedChanged

        If myDoNothing Then Return

        If optYesterday.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.YesterdayOrLastWorkingDay, Me.LastWorkingday)
        ElseIf optFromStartOfCurrentMonthToNow.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.FromStartOfCurrentMonthToNow)
        ElseIf optFromStartOfCurrentWeekToNow.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.FromStartOfCurrentWeekToNow)
        ElseIf optStartToEndOfSpecifiedMonth.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, cmbMonthsHistory.SelectedIndex)
        ElseIf optLastWeek.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.LastWeek)
        ElseIf optToday.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.Today)
        ElseIf optSinceYearBegan.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.SinceYearBeganToNow)
        ElseIf optWeekBeforeLastWeek.Checked Then
            DateRangeValue = New DateRangeParameter(DateRangePresets.WeekBeforeLastWeek)
        ElseIf optCustomPeriod.Checked Then
            DateRangeValue = New DateRangeParameter(dtpStart.Value, dtpEnd.Value)
        End If
    End Sub

    Private Sub cmbMonthsHistory_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmbMonthsHistory.SelectedIndexChanged
        If Not optStartToEndOfSpecifiedMonth.Checked Then
            optStartToEndOfSpecifiedMonth.Checked = True
        Else
            DateRangeValue = New DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, cmbMonthsHistory.SelectedIndex)
        End If
    End Sub

    Private Sub gbTitle_Enter(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles gbTitle.Enter

    End Sub
End Class

File: src/NetFxCSharp/Facesso.GenericControls/ucAnalysisDateRangePicker.cs
using ActiveDev;
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.GenericControls
{
    public partial class ucAnalysisDateRangePicker
    {
        private DateRangeParameter myDateRangeValue;
        private bool myDoNothing;
        private LastWorkingdays myLastWorkingday;
        public ucAnalysisDateRangePicker()
        {
            // This call is required by the Windows Form Designer.
            myDoNothing = true;
            InitializeComponent();
            myDoNothing = false;
            myLastWorkingday = LastWorkingdays.Friday;
            // Add any initialization after the InitializeComponent() call.
            cmbMonthsHistory.Items.Add("Laufender Monat (" + MonthText(0) + ")");
            cmbMonthsHistory.Items.Add("Letzter Monat (" + MonthText(1) + ")");
            for (int c = 2; c <= 24; c++)
            {
                cmbMonthsHistory.Items.Add("Vor " + c + " Monaten (" + MonthText(c) + ")");
            }

            DateRangeValue = new DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, 1);
            myDateRangeValue.LastWorkingday = LastWorkingdays.Friday;
        }

        private string MonthText(int MonthsIntoPast)
        {
            System.DateTime locStartDate = default(System.DateTime);
            locStartDate = Dates.LastDayOfMonth(System.DateTime.Now).AddMonths(-MonthsIntoPast);
            return locStartDate.ToString("MMM, yyyy");
        }

        public DateRangeParameter DateRangeValue
        {
            get
            {
                return myDateRangeValue;
            }

            set
            {
                myDateRangeValue = value;
                UpdateUI();
            }
        }

        public LastWorkingdays LastWorkingday
        {
            get
            {
                return myLastWorkingday;
            }

            set
            {
                myLastWorkingday = value;
                myDateRangeValue.LastWorkingday = value;
            }
        }

        private void UpdateUI()
        {
            myDoNothing = true;
            {
                var __select0 = (int)(myDateRangeValue.DateRangePreset);
                if (__select0 == (int)(DateRangePresets.CustomPeriod))
                {
                    optCustomPeriod.Checked = true;
                    myDoNothing = false;
                    return;
                }
                else if (__select0 == (int)(DateRangePresets.YesterdayOrLastWorkingDay))
                {
                    optYesterday.Checked = true;
                }
                else if (__select0 == (int)(DateRangePresets.FromStartOfCurrentMonthToNow))
                {
                    optFromStartOfCurrentMonthToNow.Checked = true;
                }
                else if (__select0 == (int)(DateRangePresets.FromStartOfCurrentWeekToNow))
                {
                    optFromStartOfCurrentWeekToNow.Checked = true;
                }
                else if (__select0 == (int)(DateRangePresets.FromStartToEndOfSpecifiedMonth))
                {
                    cmbMonthsHistory.SelectedIndex = DateRangeValue.MonthIntoPast;
                }
                else if (__select0 == (int)(DateRangePresets.LastWeek))
                {
                    optLastWeek.Checked = true;
                }
                else if (__select0 == (int)(DateRangePresets.Today))
                {
                    optToday.Checked = true;
                }
                else if (__select0 == (int)(DateRangePresets.SinceYearBeganToNow))
                {
                    optSinceYearBegan.Checked = true;
                }
                else
                {
                    optWeekBeforeLastWeek.Checked = true;
                }
            }

            dtpStart.Value = myDateRangeValue.StartDate;
            dtpEnd.Value = myDateRangeValue.EndDate;
            myDoNothing = false;
        }

        private void dtpStart_ValueChanged(System.Object sender, System.EventArgs e)
        {
            if (myDoNothing)
            {
                return;
            }

            if (dtpStart.Value > dtpEnd.Value)
            {
                dtpEnd.Value = dtpStart.Value;
            }

            DateRangeValue = new DateRangeParameter(dtpStart.Value, dtpEnd.Value);
        }

        private void dtpEnd_ValueChanged(System.Object sender, System.EventArgs e)
        {
            if (myDoNothing)
            {
                return;
            }

            if (dtpStart.Value > dtpEnd.Value)
            {
                dtpStart.Value = dtpEnd.Value;
            }

            DateRangeValue = new DateRangeParameter(dtpStart.Value, dtpEnd.Value);
        }

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        public override string Text
        {
            get
            {
                return base.Text;
            }

            set
            {
                base.Text = value;
            }
        }

        protected override void OnTextChanged(System.EventArgs e)
        {
            base.OnTextChanged(e);
            gbTitle.Text = this.Text;
        }

        private void optDateRanges_CheckedChanged(System.Object sender, System.EventArgs e)
        {
            if (myDoNothing)
            {
                return;
            }

            if (optYesterday.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.YesterdayOrLastWorkingDay, this.LastWorkingday);
            }
            else if (optFromStartOfCurrentMonthToNow.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.FromStartOfCurrentMonthToNow);
            }
            else if (optFromStartOfCurrentWeekToNow.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.FromStartOfCurrentWeekToNow);
            }
            else if (optStartToEndOfSpecifiedMonth.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, cmbMonthsHistory.SelectedIndex);
            }
            else if (optLastWeek.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.LastWeek);
            }
            else if (optToday.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.Today);
            }
            else if (optSinceYearBegan.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.SinceYearBeganToNow);
            }
            else if (optWeekBeforeLastWeek.Checked)
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.WeekBeforeLastWeek);
            }
            else if (optCustomPeriod.Checked)
            {
                DateRangeValue = new DateRangeParameter(dtpStart.Value, dtpEnd.Value);
            }
        }

        private void cmbMonthsHistory_SelectedIndexChanged(System.Object sender, System.EventArgs e)
        {
            if (!(optStartToEndOfSpecifiedMonth.Checked))
            {
                optStartToEndOfSpecifiedMonth.Checked = true;
            }
            else
            {
                DateRangeValue = new DateRangeParameter(DateRangePresets.FromStartToEndOfSpecifiedMonth, cmbMonthsHistory.SelectedIndex);
            }
        }

        private void gbTitle_Enter(System.Object sender, System.EventArgs e)
        {
        }
    }
}